src/bin/pg_autoctl/cli_do_tmux.c

```diff
@@ -439,7 +514,333 @@ cli_do_tmux_script(int argc, char **argv)
+		log_warn("Failed to kill tmux sessions \"%s\"", sessionName);
+	}
+
+	free_program(&program);
+
+	return success;
+}
+
+
+/*
+ * keeper_cli_tmux_script generates a tmux script to run a test case or a demo
+ * for pg_auto_failover easily.
+ */
+void
+cli_do_tmux_script(int argc, char **argv)
+{
+	TmuxOptions options = tmuxOptions;
+	PQExpBuffer script = createPQExpBuffer();
+
+	if (script == NULL)
+	{
+		log_error("Failed to allocate memory");
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/* prepare the tmux script */
+	(void) prepare_tmux_script(&options, script, true);
+
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(script))
+	{
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(script);
+
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	fformat(stdout, "%s", script->data);
+
+	destroyPQExpBuffer(script);
+}
+
+
+/*
+ * cli_do_tmux_session starts an interactive tmux session with the given
+ * specifications for a cluster. When the session is detached, the pg_autoctl
+ * processes are stopped.
+ */
+void
+cli_do_tmux_session(int argc, char **argv)
+{
+	TmuxOptions options = tmuxOptions;
+	PQExpBuffer script = createPQExpBuffer();
+
+	char scriptName[MAXPGPATH] = { 0 };
+
+	bool success = true;
+
+	/*
+	 * Write the script to "script-${first-pgport}.tmux" file in the root
+	 * directory.
+	 */
+	log_debug("mkdir -p \"%s\"", options.root);
+	if (!ensure_empty_dir(options.root, 0700))
+	{
+		/* errors have already been logged. */
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	if (!normalize_filename(options.root, options.root, sizeof(options.root)))
+	{
+		/* errors have already been logged. */
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/*
+	 * Prepare the tmux script.
+	 */
+	if (script == NULL)
+	{
+		log_error("Failed to allocate memory");
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	(void) prepare_tmux_script(&options, script, false);
+
+	/* memory allocation could have failed while building string */
+	if (PQExpBufferBroken(script))
+	{
+		log_error("Failed to allocate memory");
+		destroyPQExpBuffer(script);
+
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	/*
+	 * Write the script to file.
+	 */
+	sformat(scriptName, sizeof(scriptName),
+			"%s/script-%d.tmux", options.root, options.firstPort);
+
+	log_info("Writing tmux session script \"%s\"", scriptName);
+
+	if (!write_file(script->data, script->len, scriptName))
+	{
+		log_fatal("Failed to write tmux script at \"%s\"", scriptName);
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	destroyPQExpBuffer(script);
+
+	/*
+	 * Start a tmux session from the script.
+	 */
+	if (!tmux_start_server(options.root, scriptName))
+	{
+		success = false;
+		log_fatal("Failed to start the tmux session, see above for details");
+	}
+
+	/*
+	 * Stop our pg_autoctl processes and kill the tmux session.
+	 */
+	log_info("tmux session ended: kill pg_autoct processes");
+	success = success && tmux_stop_pg_autoctl(&options);
+	success = success && tmux_kill_session(&options);
```

Comment:
Maybe it makes more sense to kill the session first. Otherwise it might still spawn a `pg_autoctl` process in between this process trying to kill it, but before the `tmux` session is killed.

Reply:
I'm still not sure how that works, the kill session thing, with respect to processes running in the shell in the panes. That's why I want to first stop our processes and then kill the session. Does it make sense?